docassemble/ALWeaver/data/templates/output.mako

```diff
@@ -108,7 +94,9 @@ id: interview_order_${ interview.interview_label }
 code: |
   % if generate_download_screen:
   # Set the allowed courts for this interview
-  allowed_courts = interview_metadata["${ interview.interview_label }"]["allowed courts"]
+  % if interview.court_related:
+  allowed_courts = ${ repr(sorted(interview.allowed_courts.true_values() + (interview.allowed_courts_text.split(",") if interview.allowed_courts.get("Other") else []))) }
```

Comment:
This could still come from the metadata block, using `all_variables(special='metadata')`, right? That would change if someone makes this a nested interview, but so do a lot of other things. And, TBH it makes the most sense to only take allowed courts from the parent interview, so keeping this line as is would be fine, IMO.

Reply:
It's possible for multiple interviews in a form set to have a different set of allowed courts--that's true for the 209A packet, for example. From memory, that's why we defaulted to setting the allowed_courts in the interview order block instead of using a different pattern, like a named block.

Original code was a bit circuitous in arriving at that state, but the reference to `interview_metadata` meant it was always pulling from the "local" form's set of allowed courts anyway.

There will be circumstances where a different choice makes sense, but that's something authors can still control. Going to keep this change as-is.

Reply:
> It's possible for multiple interviews in a form set to have a different set of allowed courts

Yes, and that's fine; but what does the umbrella interview advertise as its allowed courts then? This is something we'll need to figure out and handle if we want to address #169 ever. I'm also finding it helpful to work through e-filing situations too.  The cases I can think of:

1. sub-interview has a strict superset of courts of the main interview (i.e. anything and the Affidavit; the affidavit can be filed anywhere, and the main interview has a smaller set of courts it can be filed in). In this case, the allowed courts should just be taken from the main interview, whatever that may be (what I was suggesting). 
2. sub-interview has a strict subset of courts to the main interview; for example, the main interview can be filed in Housing and Probate, but the sub-interview can only be filed in Housing.
    * If the sub-interview is required, than the main interview isn't really able to be filed in Housing and probate, right? That doesn't really make sense and is invalid, so the main interview should change the allowed courts. 
    * If the sub-interview isn't required, then the allowed courts actually change based on whether or not someone is using the sub-interview; IMO in this case, the advertised allowed courts should stay the same, but the logic of what the allowed courts actually is will need to be gets more complicated, and we can't really represent that with the metadata block
3. sub-interview has at least one court in common with the main interview. This is some combination of 1) and 2.2), which sorta defaults to the more complicated logic case of 2.2.
4. sub-interview needs to be filed completely separately to a different court from the main interview (like the Notice of Appeal in the Motion to Stay eviction, which needs to go to the housing court, not the Appeals court). Is this the case you're thinking of that we would want to avoid getting the info from the parent's metadata block? It makes sense as a use case, but everything else in the Weaver and e-filing are assuming that things go to the same court (that's why we have a singular `al_court_bundle`, vs the user bundle, who have to manually file to those other courts at the moment).
 
 4 is kinda confusing to me, but IMO for what the weaver can do now, importing from the parent's / main interview metadata block is a good default. I do think that the duplicate of `allowed_courts` like this will result in the metadata block not getting updated after going through the Weaver ever.

Reply:
It's not really a mechanical process like you're describing. In the 209A, we offer all courts, then ask the litigant to confirm their choice if they select a kind of relief that's only available in a narrower set of courts. Specifically in the 209A, the user might prefer a District or Municipal Court for safety and comfort, but if they want any orders relating to children they need to file in Probate Court. The confirmation/rejection allows the user to change their mind. The logic is human based, depending on the needs of the litigant and we clearly explain the tradeoffs. The logic will be like this in most interviews, hardcoded to a specific combination of forms. We'll never want automatic logic that combines the various sets of courts and comes up with some subset.

When we used the `interview_metadata` block, at least it had the benefit of showing the list of courts that was listed in the same YAML file, but it was still kind of obscure and led to difficult to trace problems when you included one YAML in another. If we use the "last included" metadata block to set the list of courts, there's a larger risk of hard to track down errors. For a specific recent experience, it will be very hard to see what is setting the list of courts in the "show variables and values" debug scenario. Ran into this recently with a child interview doing this in UpToCode.

Having done a few "combined" interviews, I don't think #169 makes much sense to be part of the Weaver, but I'm still willing to see what happens with it. I'm not convinced enough to plan other features around it.

I can see that the situation might be a little different for e-filing. But for some of the reasons you've pointed out, the `metadata` block can be an unreliable source to get the list of courts from. It seems to me that an "allowed_courts" code block also may not work for all e-filing scenarios, as you want a reliable source of information on a per-document basis, while the "allowed_courts" variable is used to show the user the right question and isn't tied to a document at all.

Should there be an e-filing specific block to refer to? It can be nice to double-up uses of similar data, but as we're seeing here, the needs don't always line up. Honestly, the needs of an e-filing interview are pretty different. One interview might e-file to two courts. One "child interview" might assemble 2 documents. It would probably make more sense to have this metadata tied to the ALDocument than the interview metadata.

I'm also open to moving the `allowed_courts` definition to a standalone code block and removing it from the metadata block. I don't think that solves the e-filing need at all, but it does mean there's one place to look and it's clearly visible in the `show variables and values` trace and in the playground's list of variable definitions. It also has the benefit of using Docassemble's normal priority system to determine which interview "wins" the decision of setting the list of allowed courts.

Perhaps that's actually the best solution.

Reply:
Thanks for going in depth with me on this btw, I appreciate it, I get really lost where our forms hit IRL court processes.

> Having done a few "combined" interviews, I don't think #169 makes much sense to be part of the Weaver, but I'm still willing to see what happens with it. I'm not convinced enough to plan other features around it.

That's fair. It is probably a lot of effort to get that feature to a point where it'd be really useful. I kept it open because I don't think it's solved, but saying that I'd try it out when I was done with e-filing was, well, ambitious 😅 . Happy to close it as not completed.

> If we use the "last included" metadata block to set the list of courts, there's a larger risk of hard to track down errors

This might be a key misunderstanding; pretty sure there should only ever be one metadata block in each interview. It'd go in the main interview, i.e. the one with the `mandatory: True` main order, and if we split something into a `body.yml`, the body doesn't get this metadata block. I agree that trying to emulate having multiple metadata blocks with `interview_metadata` did kinda make this confusing, but IMO it'd be a lot simpler with just the one `metadata` block.

> Specifically in the 209A, the user might prefer a District or Municipal Court for safety and comfort, but if they want any orders relating to children they need to file in Probate Court. The confirmation/rejection allows the user to change their mind.

This is actually convincing me that we should still take `allowed_courts` from metadata. I'm understanding that we need separate features for lists of courts:
1. the metadata block is for advertising to another website what courts that form is for. It isn't binding, but should be accurate and include all possibilities.
2. The code block then directly sets the choices people have for where they want to use the form, likely near the beginning of the interview.
3. some (potentially multiple, usually only one) checks in the interview that determines "can you actually file in this court with your desired protections, selected options, etc.", and explains if you can't.

IMO the Weaver should help with the default case, and in the _default_ case of only one main interview (if we close #169), 1 and 2 should be the same list, and if 3 is necessary, it should be hand coded.

>  It would probably make more sense to have this metadata tied to the ALDocument than the interview metadata.

TBF, that is what we're currently doing. I'm just not super happy with it, and am trying to think of other places that could help. But you've convinced me about the separate uses, and e-filing has a lot of weird API quirks anyway, so this probably isn't a good place for it.

Reply:
The main challenge with taking the list of courts from the metadata is that it's hard to understand what's going on. 

In Docassemble, only one metadata block will "win" but if we do a lookup using `all_variables(special='metadata')`, you can't easily inspect the value of the "winning" metadata block. The source is not displayed in `show variables and values`. The list of courts that wins wouldn't be visible in the debug trace like a normal code block would be. The only way to check would be to reason through the order of inclusion of each YAML. Agreed that this is a good reasons to have only one metadata block, but by definition this is something we would encounter when tracing unexpected behavior.

Is the list of allowed courts really metadata about the interview, or is it primarily a substantive value? Is the idea that we display this on courtformsonline.org? I can see some value in displaying it somewhere, but I'm not sure it will be as informative to litigants as the topic of the form (especially as the concept of state-wide courts with limited substantive jurisdiction is Massachusetts-specific). In addition, while it's fine to set as a static choice in most interviews, there are some where having it happen as a code block is required. The list of metadata is too restrictive to handle conditional court options.

I suppose if we really do want it to pull from `metadata` we can improve debugging if the value is pulled in a standalone code block, perhaps with a comment that explains to check the value of the "winning" metadata block. That would improve it from the old method a bit. I'm just having trouble seeing the value in this less transparent approach. If we're trying to simplify and keep just one place to store the information, I would still lean to removing it from `metadata`.